pkg/interactive/docker.go

```diff
@@ -76,39 +74,36 @@ func (d *Docker) AddMount(typeStr, source, target string, optAdditionalArgs ...s
 	d.AddArgs("--mount", fmt.Sprintf("type=%s,source=%s,target=%s%s", typeStr, source, target, addl))
 }
 
-// AddRWOverlay mounts a directory into the image at the desired location, but with an overlay
-//  so internal changes do not modify the external directory.
-// externalDirectory probably needs to be an absolute path
-// Returns a function to clean up the mount (but does not delete the directory).
-// Uses sudo and probably only works on Linux
-func (d *Docker) AddRWOverlay(externalDirectory, internalDirectory string) func() {
-	tmpDir, err := ioutil.TempDir("", "overlay")
+// CopyAndAddMount copies the source files into a temp directory, and then
+// mounts them as the target. It also returns a function to remove the temp
+// directory for cleaning up.
+func (d *Docker) CopyAndAddMount(typeStr, parentDir, source, target string, optAdditionalArgs ...string) (func(), error) {
+	fi, err := os.Stat(source)
 	if err != nil {
-		log.Fatal(err)
+		return nil, fmt.Errorf("error getting the FileInfo for %q: %w", source, err)
 	}
-	subDirs := []string{"upper", "work", "overlay"}
-	for _, d := range subDirs {
-		err = os.Mkdir(path.Join(tmpDir, d), 0777)
-		if err != nil {
-			log.Fatal(err)
-		}
+
+	tempDir, err := ioutil.TempDir(parentDir, fi.Name())
+	if err != nil {
+		return nil, fmt.Errorf("error creating a temporary directory: %w", err)
+	}
+	if err := os.Chmod(tempDir, 0777); err != nil {
+		return nil, fmt.Errorf("error changing file mode for %q: %w", tempDir, err)
 	}
-	overlayDir := path.Join(tmpDir, "overlay")
-	// The options for overlay mount are confusing
-	// You need empty directories for upper and work (and overlay, though you can mount over directories that have files in them if you *want* to...)
-	mount := NewCommand("sudo", "mount", "-t", "overlay", "-o",
-		fmt.Sprintf("lowerdir=%s,upperdir=%s/upper,workdir=%s/work", externalDirectory, tmpDir, tmpDir),
-		"none", overlayDir)
-	// Print command to run so user knows why it is asking for sudo password (if it does)
-	log.Println(mount)
-	if err = mount.Run(); err != nil {
-		log.Fatalf("Unable to create overlay mount, so giving up: %v", err)
+	cleanup := func() {
+		os.RemoveAll(tempDir)
```

Comment:
Can we have some error checking control done here with a log statement?

Reply:
Added some checking control here, but since it'll be a best-effort cleanup so its failure should not impact the final results, I'm using `warning` instead of `fatalf` log here.